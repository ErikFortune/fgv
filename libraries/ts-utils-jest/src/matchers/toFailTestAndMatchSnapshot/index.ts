/* eslint-disable @typescript-eslint/no-unused-vars,no-unused-vars */
import { Context, toMatchSnapshot } from 'jest-snapshot';
import { matcherName, predicate } from './predicate';

import { matcherHint } from 'jest-matcher-utils';
import { stripAnsiColors } from '../../utils/colorHelpers';

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace jest {
    // eslint-disable-next-line no-unused-vars, @typescript-eslint/no-unused-vars,  @typescript-eslint/naming-convention
    interface Matchers<R, T> {
      /**
       * Use .toFailTestWith to test a custom matcher by
       * verifying that a test case fails as expected and
       * reports an error matching a stored snapshot.
       */
      toFailTestAndMatchSnapshot<T>(): R;
    }
  }
}

export default {
  toFailTestAndMatchSnapshot: function <T>(
    this: jest.MatcherContext,
    cb: () => void
  ): jest.CustomMatcherResult {
    const context = this as unknown as Context;
    const cbResult = predicate(cb);
    if (cbResult.isFailure()) {
      return {
        pass: false,
        message: (): string => {
          return [
            matcherHint(`${matcherName}`, 'callback'),
            '  Expected: Callback to fail with an error that matches snapshot',
            '  Received: Callback succeeded'
          ].join('\n');
        }
      };
    }
    return toMatchSnapshot.call(
      context,
      stripAnsiColors(cbResult.value),
      'toFailTestAndMatchSnapshot'
    ) as jest.CustomMatcherResult;
  }
};
