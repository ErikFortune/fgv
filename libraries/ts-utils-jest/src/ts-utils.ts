export * from '@fgv/ts-utils';
