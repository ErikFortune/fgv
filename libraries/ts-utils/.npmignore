# Agent-specific files
CLAUDE.md
.agents/
.claude/

# IDE and system files
.vscode/
.DS_Store

# Test and coverage files  
coverage/
*.test.ts
*.test.js
*.test.d.ts
*.test.*.map
src/test/
lib/test/
lib/**/*.test.*
lib-commonjs/test/
lib-commonjs/**/*.test.*

# Development files
*.log
.eslintrc*
jest.config.*
tsconfig.json
*.map

# Source files (if publishing only built output)
# Uncomment if you only want to publish dist/lib
src/

# Rush/Heft temporary files
.heft/
.rush/
rush-logs/
temp/

# API Extractor files (development artifacts)
etc/

# Build artifacts that shouldn't be published
dist/
lib-commonjs/

# Documentation source
docs/
*.md
!README.md
!LICENSE.md