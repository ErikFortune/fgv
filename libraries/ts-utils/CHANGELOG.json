{
  "name": "@fgv/ts-utils",
  "entries": [
    {
      "version": "4.2.0",
      "tag": "@fgv/ts-utils_v4.2.0",
      "date": "Mon, 20 Jan 2025 09:46:53 GMT",
      "comments": {
        "none": [
          {
            "comment": "add collections packlet and ResultMap class"
          },
          {
            "comment": "add withFormattedError on validator and converter"
          },
          {
            "comment": "clean up type inconsistencies, relax some constraints"
          }
        ]
      }
    },
    {
      "version": "4.1.0",
      "tag": "@fgv/ts-utils_v4.1.0",
      "date": "Thu, 09 Jan 2025 05:33:39 GMT",
      "comments": {
        "none": [
          {
            "comment": "update dependencies"
          }
        ]
      }
    },
    {
      "version": "4.0.2",
      "tag": "@fgv/ts-utils_v4.0.2",
      "date": "Tue, 14 May 2024 14:45:53 GMT",
      "comments": {}
    },
    {
      "version": "4.0.1",
      "tag": "@fgv/ts-utils_v4.0.1",
      "date": "Tue, 14 May 2024 05:02:20 GMT",
      "comments": {
        "none": [
          {
            "comment": "publish"
          },
          {
            "comment": "import TextEncoder to avoid globals race"
          }
        ]
      }
    },
    {
      "version": "4.0.0",
      "tag": "@fgv/ts-utils_v4.0.0",
      "date": "Tue, 14 May 2024 03:09:27 GMT",
      "comments": {
        "none": [
          {
            "comment": "rework validation/converter compatibility a bit"
          },
          {
            "comment": "QoL improvement to simplify error aggregation"
          },
          {
            "comment": "factor optional converters out to ts-extras package to eliminate large dependencies"
          },
          {
            "comment": "update generated api docs"
          },
          {
            "comment": "add convalidate helper"
          },
          {
            "comment": "add defaulting converter"
          },
          {
            "comment": "extend conversion to accept validators"
          },
          {
            "comment": "export StringConverter at top level"
          }
        ]
      }
    },
    {
      "version": "3.0.0",
      "tag": "@fgv/ts-utils_v3.0.0",
      "date": "Mon, 22 Jan 2024 07:00:18 GMT",
      "comments": {
        "none": [
          {
            "comment": "gitignore tweaks"
          },
          {
            "comment": "refactor hash implementation"
          },
          {
            "comment": "bump versions"
          },
          {
            "comment": "refactor and cleanup"
          },
          {
            "comment": "build changes"
          },
          {
            "comment": "Factor out extras, enable shaking"
          },
          {
            "comment": "bump dependencies"
          },
          {
            "comment": "version upgrades"
          },
          {
            "comment": "relax mapToRecord from Map to ReadonlyMap"
          },
          {
            "comment": "bump version"
          },
          {
            "comment": "add oneOf validator"
          }
        ]
      }
    },
    {
      "version": "2.0.0",
      "tag": "@fgv/ts-utils_v2.0.0",
      "date": "Mon, 10 Apr 2023 03:33:11 GMT",
      "comments": {
        "major": [
          {
            "comment": "refactor and cleanup"
          },
          {
            "comment": "reorganization and cleanup"
          },
          {
            "comment": "more refactoring"
          },
          {
            "comment": "refactor"
          }
        ]
      }
    }
  ]
}
