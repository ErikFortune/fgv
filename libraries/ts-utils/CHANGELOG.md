# Change Log - @fgv/ts-utils

This log was last generated on Thu, 30 Jan 2025 00:35:17 GMT and should not be manually modified.

## 4.3.0
Thu, 30 Jan 2025 00:35:17 GMT

### Updates

- add collector, convertingResultMap, convertingCollector

## 4.2.2
Thu, 23 Jan 2025 06:19:32 GMT

### Updates

- QoL improvements to Success/Failure

## 4.2.1
Tue, 21 Jan 2025 04:19:21 GMT

### Updates

- allow factories in getOrAdd

## 4.2.0
Mon, 20 Jan 2025 09:46:53 GMT

### Updates

- add collections packlet and ResultMap class
- add withFormattedError on validator and converter
- clean up type inconsistencies, relax some constraints

## 4.1.0
Thu, 09 Jan 2025 05:33:39 GMT

### Updates

- update dependencies

## 4.0.2
Tue, 14 May 2024 14:45:53 GMT

_Version update only_

## 4.0.1
Tue, 14 May 2024 05:02:20 GMT

### Updates

- publish
- import TextEncoder to avoid globals race

## 4.0.0
Tue, 14 May 2024 03:09:27 GMT

### Updates

- rework validation/converter compatibility a bit
- QoL improvement to simplify error aggregation
- factor optional converters out to ts-extras package to eliminate large dependencies
- update generated api docs
- add convalidate helper
- add defaulting converter
- extend conversion to accept validators
- export StringConverter at top level

## 3.0.0
Mon, 22 Jan 2024 07:00:18 GMT

### Updates

- gitignore tweaks
- refactor hash implementation
- bump versions
- refactor and cleanup
- build changes
- Factor out extras, enable shaking
- bump dependencies
- version upgrades
- relax mapToRecord from Map to ReadonlyMap
- bump version
- add oneOf validator

## 2.0.0
Mon, 10 Apr 2023 03:33:11 GMT

### Breaking changes

- refactor and cleanup
- reorganization and cleanup
- more refactoring
- refactor

