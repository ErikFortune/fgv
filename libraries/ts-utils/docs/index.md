<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md)

## API Reference

## Packages

<table><thead><tr><th>

Package


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[@fgv/ts-utils](./ts-utils.md)


</td><td>


</td></tr>
</tbody></table>
