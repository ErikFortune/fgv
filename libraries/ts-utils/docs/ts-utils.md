<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md) &gt; [@fgv/ts-utils](./ts-utils.md)

## ts-utils package

## Classes

<table><thead><tr><th>

Class


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[Collector](./ts-utils.collector.md)


</td><td>

A [Collector](./ts-utils.collections.collector.md) that is a specialized collection which contains items of type [ICollectible](./ts-utils.collections.icollectible.md)<!-- -->, which have a unique key and a write-once index.

Items are assigned an index sequentially as they are added to the collection. Once added, items are immutable - they cannot be removed or replaced.


</td></tr>
<tr><td>

[ConvertingCollector](./ts-utils.convertingcollector.md)


</td><td>

A [collector](./ts-utils.collector.md) that collects [ICollectible](./ts-utils.collections.icollectible.md) items, optionally converting them from a source representation to the target representation using a factory supplied at default or at the time of collection.


</td></tr>
<tr><td>

[DetailedFailure](./ts-utils.detailedfailure.md)


</td><td>

A [DetailedFailure&lt;T, TD&gt;](./ts-utils.detailedfailure.md) extends [Failure&lt;T&gt;](./ts-utils.failure.md) to report optional failure details in addition to the error message.


</td></tr>
<tr><td>

[DetailedSuccess](./ts-utils.detailedsuccess.md)


</td><td>

A [DetailedSuccess](./ts-utils.detailedsuccess.md) extends [Success](./ts-utils.success.md) to report optional success details in addition to the error message.


</td></tr>
<tr><td>

[Failure](./ts-utils.failure.md)


</td><td>

Reports a failed [result](./ts-utils.iresult.md) from some operation, with an error message.


</td></tr>
<tr><td>

[MessageAggregator](./ts-utils.messageaggregator.md)


</td><td>

A simple error aggregator to simplify collecting and reporting all errors in a flow.


</td></tr>
<tr><td>

[Normalizer](./ts-utils.normalizer.md)


</td><td>

Normalizes an arbitrary JSON object


</td></tr>
<tr><td>

[ObjectConverter](./ts-utils.objectconverter.md)


</td><td>

A [Converter](./ts-utils.converter.md) which converts an object of type `<T>` without changing shape, given a [FieldConverters&lt;T&gt;](./ts-utils.conversion.fieldconverters.md) for the fields in the object.


</td></tr>
<tr><td>

[ResultMap](./ts-utils.resultmap.md)


</td><td>

A [ResultMap](./ts-utils.collections.resultmap.md) class as a `Map<TK, TV>`<!-- -->-like object which reports success or failure with additional details using the [result pattern](https://github.com/ErikFortune/fgv/tree/main/libraries/ts-utils#the-result-pattern)<!-- -->.


</td></tr>
<tr><td>

[StringConverter](./ts-utils.stringconverter.md)


</td><td>

The [StringConverter](./ts-utils.conversion.stringconverter.md) class extends [BaseConverter](./ts-utils.conversion.baseconverter.md) to provide string-specific helper methods.


</td></tr>
<tr><td>

[Success](./ts-utils.success.md)


</td><td>

Reports a successful [result](./ts-utils.iresult.md) from some operation and the corresponding value.


</td></tr>
<tr><td>

[ValidatingCollector](./ts-utils.validatingcollector.md)


</td><td>

A [Collector](./ts-utils.collections.collector.md) with a [CollectorValidator](./ts-utils.collections.collectorvalidator.md) property that enables validated use of the underlying map with weakly-typed keys and values.


</td></tr>
<tr><td>

[ValidatingConvertingCollector](./ts-utils.validatingconvertingcollector.md)


</td><td>

A [ConvertingCollector](./ts-utils.collections.convertingcollector.md) with a [ConvertingCollectorValidator](./ts-utils.collections.convertingcollectorvalidator.md) property that enables validated use of the underlying map with weakly-typed keys and values.


</td></tr>
<tr><td>

[ValidatingResultMap](./ts-utils.validatingresultmap.md)


</td><td>

A [ResultMap](./ts-utils.collections.resultmap.md) with a [validator](./ts-utils.collections.resultmapvalidator.md) property that enables validated use of the underlying map with weakly-typed keys and values.


</td></tr>
</tbody></table>

## Functions

<table><thead><tr><th>

Function


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[allSucceed(results, successValue, aggregatedErrors)](./ts-utils.allsucceed.md)


</td><td>

Determines if an iterable collection of [Result&lt;T&gt;](./ts-utils.result.md) were all successful.


</td></tr>
<tr><td>

[captureResult(func)](./ts-utils.captureresult.md)


</td><td>

Wraps a function which might throw to convert exception results to [Failure](./ts-utils.failure.md)<!-- -->.


</td></tr>
<tr><td>

[fail\_2(message)](./ts-utils.fail_2.md)


</td><td>

Returns [Failure&lt;T&gt;](./ts-utils.failure.md) with the supplied error message.


</td></tr>
<tr><td>

[fails(message)](./ts-utils.fails.md)


</td><td>

Returns [Failure&lt;T&gt;](./ts-utils.failure.md) with the supplied error message.


</td></tr>
<tr><td>

[failsWithDetail(message, detail)](./ts-utils.failswithdetail.md)


</td><td>

Returns [DetailedFailure&lt;T, TD&gt;](./ts-utils.detailedfailure.md) with a supplied error message and detail.


</td></tr>
<tr><td>

[failWithDetail(message, detail)](./ts-utils.failwithdetail.md)


</td><td>

Returns [DetailedFailure&lt;T, TD&gt;](./ts-utils.detailedfailure.md) with a supplied error message and detail.


</td></tr>
<tr><td>

[firstSuccess(results)](./ts-utils.firstsuccess.md)


</td><td>

Returns the first successful result from a collection of [Result&lt;T&gt;](./ts-utils.result.md) or [DeferredResult&lt;T&gt;](./ts-utils.deferredresult.md)<!-- -->.


</td></tr>
<tr><td>

[getTypeOfProperty(key, item)](./ts-utils.gettypeofproperty.md)


</td><td>

Gets the type of a property specified by key from an arbitrary object.


</td></tr>
<tr><td>

[getValueOfPropertyOrDefault(key, item, defaultValue)](./ts-utils.getvalueofpropertyordefault.md)


</td><td>

Gets the value of a property specified by key from an arbitrary object, or a default value if the property does not exist.


</td></tr>
<tr><td>

[isDeferredResult(result)](./ts-utils.isdeferredresult.md)


</td><td>

Checks if a result is a deferred result.


</td></tr>
<tr><td>

[isKeyOf(key, item)](./ts-utils.iskeyof.md)


</td><td>

Helper type-guard function to report whether a specified key is present in a supplied object.


</td></tr>
<tr><td>

[mapDetailedResults(results, ignore, aggregatedErrors)](./ts-utils.mapdetailedresults.md)


</td><td>

Aggregates successful results from a collection of [DetailedResult&lt;T, TD&gt;](./ts-utils.detailedresult.md)<!-- -->, optionally ignoring certain error details.


</td></tr>
<tr><td>

[mapFailures(results, aggregatedErrors)](./ts-utils.mapfailures.md)


</td><td>

Aggregates error messages from a collection of [Result&lt;T&gt;](./ts-utils.result.md)<!-- -->.


</td></tr>
<tr><td>

[mapResults(results, aggregatedErrors)](./ts-utils.mapresults.md)


</td><td>

Aggregates successful result values from a collection of [Result&lt;T&gt;](./ts-utils.result.md)<!-- -->.


</td></tr>
<tr><td>

[mapSuccess(results, aggregatedErrors)](./ts-utils.mapsuccess.md)


</td><td>

Aggregates successful results from a a collection of [Result&lt;T&gt;](./ts-utils.result.md)<!-- -->.


</td></tr>
<tr><td>

[mapToRecord(src, factory)](./ts-utils.maptorecord.md)


</td><td>

Applies a factory method to convert a `ReadonlyMap<TK, TS>` into a `Record<TK, TD>`<!-- -->.


</td></tr>
<tr><td>

[omit(from, exclude)](./ts-utils.omit.md)


</td><td>

Simple implicit omit function, which picks all of the properties from a supplied object except those specified for exclusion.


</td></tr>
<tr><td>

[optionalMapToPossiblyEmptyRecord(src, factory)](./ts-utils.optionalmaptopossiblyemptyrecord.md)


</td><td>

Applies a factory method to convert an optional `ReadonlyMap<string, TS>` into a `Record<string, TD>`


</td></tr>
<tr><td>

[optionalMapToRecord(src, factory)](./ts-utils.optionalmaptorecord.md)


</td><td>

Applies a factory method to convert an optional `ReadonlyMap<string, TS>` into a `Record<string, TD>` or `undefined`<!-- -->.


</td></tr>
<tr><td>

[optionalRecordToMap(src, factory)](./ts-utils.optionalrecordtomap.md)


</td><td>

Applies a factory method to convert an optional `Record<TK, TS>` into a `Map<TK, TD>`<!-- -->, or `undefined`<!-- -->.


</td></tr>
<tr><td>

[optionalRecordToPossiblyEmptyMap(src, factory)](./ts-utils.optionalrecordtopossiblyemptymap.md)


</td><td>

Applies a factory method to convert an optional `Record<TK, TS>` into a `Map<TK, TD>`


</td></tr>
<tr><td>

[pick(from, include)](./ts-utils.pick.md)


</td><td>

Simple implicit pick function, which picks a set of properties from a supplied object. Ignores picked properties that do not exist regardless of type signature.


</td></tr>
<tr><td>

[populateObject(initializers, options, aggregatedErrors)](./ts-utils.populateobject.md)


</td><td>

Populates an an object based on a prototype full of field initializers that return [Result&lt;T\[key\]&gt;](./ts-utils.result.md)<!-- -->. Returns [Success](./ts-utils.success.md) with the populated object if all initializers succeed, or [Failure](./ts-utils.failure.md) with a concatenated list of all error messages.


</td></tr>
<tr><td>

[populateObject(initializers, order, aggregatedErrors)](./ts-utils.populateobject_1.md)


</td><td>

Populates an an object based on a prototype full of field initializers that return [Result&lt;T\[key\]&gt;](./ts-utils.result.md)<!-- -->. Returns [Success](./ts-utils.success.md) with the populated object if all initializers succeed, or [Failure](./ts-utils.failure.md) with a concatenated list of all error messages.


</td></tr>
<tr><td>

[propagateWithDetail(result, detail, successDetail)](./ts-utils.propagatewithdetail.md)


</td><td>

Propagates a [Success](./ts-utils.success.md) or [Failure](./ts-utils.failure.md) [Result](./ts-utils.result.md)<!-- -->, adding supplied event details as appropriate.


</td></tr>
<tr><td>

[recordToMap(src, factory)](./ts-utils.recordtomap.md)


</td><td>

Applies a factory method to convert a `Record<TK, TS>` into a `Map<TK, TD>`<!-- -->.


</td></tr>
<tr><td>

[succeed(value)](./ts-utils.succeed.md)


</td><td>

Returns [Success&lt;T&gt;](./ts-utils.success.md) with the supplied result value.


</td></tr>
<tr><td>

[succeeds(value)](./ts-utils.succeeds.md)


</td><td>

Returns [Success&lt;T&gt;](./ts-utils.success.md) with the supplied result value.


</td></tr>
<tr><td>

[succeedsWithDetail(value, detail)](./ts-utils.succeedswithdetail.md)


</td><td>

Returns [DetailedSuccess&lt;T, TD&gt;](./ts-utils.detailedsuccess.md) with a supplied value and optional detail.


</td></tr>
<tr><td>

[succeedWithDetail(value, detail)](./ts-utils.succeedwithdetail.md)


</td><td>

Returns [DetailedSuccess&lt;T, TD&gt;](./ts-utils.detailedsuccess.md) with a supplied value and optional detail.


</td></tr>
<tr><td>

[useOrInitialize(value, initializer)](./ts-utils.useorinitialize.md)


</td><td>

Uses a value or calls a supplied initializer if the supplied value is undefined.


</td></tr>
</tbody></table>

## Interfaces

<table><thead><tr><th>

Interface


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[Converter](./ts-utils.converter.md)


</td><td>

Generic converter to convert unknown to a templated type `<T>`<!-- -->, using intrinsic rules or as modified by an optional conversion context of optional templated type `<TC>` (default `undefined`<!-- -->).


</td></tr>
<tr><td>

[ICollectible](./ts-utils.icollectible.md)


</td><td>

An item that can be collected by some [Collector](./ts-utils.convertingcollector.md)<!-- -->.


</td></tr>
<tr><td>

[IMessageAggregator](./ts-utils.imessageaggregator.md)


</td><td>

Simple error aggregator to simplify collecting all errors in a flow.


</td></tr>
<tr><td>

[IMessageReportDetail](./ts-utils.imessagereportdetail.md)


</td><td>

Details for reporting a message.


</td></tr>
<tr><td>

[IReadOnlyResultMap](./ts-utils.ireadonlyresultmap.md)


</td><td>

A readonly `ReadonlyMap<TK, TV>`<!-- -->-like object which reports success or failure with additional details using the [result pattern](https://github.com/ErikFortune/fgv/tree/main/libraries/ts-utils#the-result-pattern)<!-- -->.


</td></tr>
<tr><td>

[IResult](./ts-utils.iresult.md)


</td><td>

Represents the result of some operation of sequence of operations.


</td></tr>
<tr><td>

[IResultLogger](./ts-utils.iresultlogger.md)


</td><td>

Simple logger interface used by [orThrow(logger)](./ts-utils.iresult.orthrow.md) and [orThrow(formatter)](./ts-utils.iresult.orthrow_1.md)<!-- -->.


</td></tr>
<tr><td>

[IResultReporter](./ts-utils.iresultreporter.md)


</td><td>

Interface for reporting a result.


</td></tr>
<tr><td>

[IResultReportOptions](./ts-utils.iresultreportoptions.md)


</td><td>

Options for reporting a result.


</td></tr>
<tr><td>

[PopulateObjectOptions](./ts-utils.populateobjectoptions.md)


</td><td>

Options for the [populateObject()](./ts-utils.populateobject.md) function.


</td></tr>
<tr><td>

[Validator](./ts-utils.validator.md)


</td><td>

In-place validation that a supplied unknown matches some required characteristics (type, values, etc).


</td></tr>
</tbody></table>

## Namespaces

<table><thead><tr><th>

Namespace


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[Collections](./ts-utils.collections.md)


</td><td>


</td></tr>
<tr><td>

[Conversion](./ts-utils.conversion.md)


</td><td>


</td></tr>
<tr><td>

[Converters](./ts-utils.converters.md)


</td><td>


</td></tr>
<tr><td>

[Hash](./ts-utils.hash.md)


</td><td>


</td></tr>
<tr><td>

[Logging](./ts-utils.logging.md)


</td><td>


</td></tr>
<tr><td>

[Validation](./ts-utils.validation.md)


</td><td>


</td></tr>
<tr><td>

[Validators](./ts-utils.validators.md)


</td><td>


</td></tr>
</tbody></table>

## Type Aliases

<table><thead><tr><th>

Type Alias


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[Brand](./ts-utils.brand.md)


</td><td>

Helper type to brand a simple type to prevent inappropriate use


</td></tr>
<tr><td>

[DeferredResult](./ts-utils.deferredresult.md)


</td><td>

Represents a deferred result that will be evaluated if needed.


</td></tr>
<tr><td>

[DetailedFailureContinuation](./ts-utils.detailedfailurecontinuation.md)


</td><td>

Callback to be called when a [DetailedResult](./ts-utils.detailedresult.md) encounters a failure.


</td></tr>
<tr><td>

[DetailedResult](./ts-utils.detailedresult.md)


</td><td>

**_(BETA)_** Type inference to determine the result type `T` of a [DetailedResult&lt;T, TD&gt;](./ts-utils.detailedresult.md)<!-- -->.


</td></tr>
<tr><td>

[DetailedSuccessContinuation](./ts-utils.detailedsuccesscontinuation.md)


</td><td>

Callback to be called when a [DetailedResult](./ts-utils.detailedresult.md) encounters success.


</td></tr>
<tr><td>

[ErrorFormatter](./ts-utils.errorformatter.md)


</td><td>

Formats an error message.


</td></tr>
<tr><td>

[FailureContinuation](./ts-utils.failurecontinuation.md)


</td><td>

Continuation callback to be called in the event that an [Result](./ts-utils.result.md) fails.


</td></tr>
<tr><td>

[FieldInitializers](./ts-utils.fieldinitializers.md)


</td><td>

String-keyed record of initialization functions to be passed to [populateObject()](./ts-utils.populateobject.md) or [populateObject()](./ts-utils.populateobject_1.md)<!-- -->.


</td></tr>
<tr><td>

[MessageLogLevel](./ts-utils.messageloglevel.md)


</td><td>

The severity level at which a message should be logged.


</td></tr>
<tr><td>

[Result](./ts-utils.result.md)


</td><td>

Represents the [result](./ts-utils.iresult.md) of some operation or sequence of operations.


</td></tr>
<tr><td>

[ResultDetailType](./ts-utils.resultdetailtype.md)


</td><td>

**_(BETA)_** Type inference to determine the detail type `TD` of a [DetailedResult&lt;T, TD&gt;](./ts-utils.detailedresult.md)<!-- -->.


</td></tr>
<tr><td>

[ResultValueType](./ts-utils.resultvaluetype.md)


</td><td>

**_(BETA)_** Type inference to determine the result type of an [Result](./ts-utils.result.md)<!-- -->.


</td></tr>
<tr><td>

[SuccessContinuation](./ts-utils.successcontinuation.md)


</td><td>

Continuation callback to be called in the event that an [Result](./ts-utils.result.md) is successful.


</td></tr>
</tbody></table>

