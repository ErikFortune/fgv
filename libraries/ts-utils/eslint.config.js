// ESLint 9 flat config for ts-utils
const nodeProfile = require('@rushstack/eslint-config/flat/profile/node');
const packletsPlugin = require('@rushstack/eslint-config/flat/mixins/packlets');
const tsdocPlugin = require('@rushstack/eslint-config/flat/mixins/tsdoc');

module.exports = [
  ...nodeProfile,
  packletsPlugin,
  ...tsdocPlugin,
  {
    // Override specific rules if needed
    rules: {
      '@rushstack/packlets/mechanics': 'warn'
    }
  }
];
