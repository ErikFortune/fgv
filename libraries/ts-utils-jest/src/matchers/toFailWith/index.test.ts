import { fail, succeed } from '../../ts-utils';
import matcher from './';
import toFailTestAndMatchSnapshot from '../toFailTestAndMatchSnapshot';

expect.extend({
  ...matcher,
  ...toFailTestAndMatchSnapshot
});

describe('.toFailWith', () => {
  test('passes with a failure result and matching string or RegExp', () => {
    expect(fail('oops')).toFailWith('oops');
    expect(fail('oops')).toFailWith(/o.*/i);
  });

  test('fails with a success result', () => {
    expect(succeed('hello')).not.toFailWith('hello');
  });

  test('fails with a failure result but non-matching string or RegExp', () => {
    expect(fail('oops')).not.toFailWith('error');
    expect(fail('oops')).not.toFailWith(/x.*/i);
  });

  test('reports details correctly on a failure due to a success result', () => {
    expect(() => {
      expect(succeed('hello')).toFailWith('hello');
    }).toFailTestAndMatchSnapshot();
  });

  test('reports details correctly on a failure due to non-matching failure value', () => {
    expect(() => {
      expect(fail('oops')).toFailWith('error');
    }).toFailTestAndMatchSnapshot();
  });

  test('reports details correctly on a failure with a .not', () => {
    expect(() => {
      expect(fail('oops')).not.toFailWith('oops');
    }).toFailTestAndMatchSnapshot();
  });
});
