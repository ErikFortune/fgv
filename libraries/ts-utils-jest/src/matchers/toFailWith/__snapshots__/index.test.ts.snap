// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.toFailWith reports details correctly on a failure due to a success result: toFailTestAndMatchSnapshot 1`] = `
"expect(received).toFailWith(expected)
Expected:
  Failure with \\"hello\\"
Received:
  Success with \\"hello\\""
`;

exports[`.toFailWith reports details correctly on a failure due to non-matching failure value: toFailTestAndMatchSnapshot 1`] = `
"expect(received).toFailWith(expected)
Expected:
  Failure with \\"error\\"
Received:
  Failure with \\"oops\\""
`;

exports[`.toFailWith reports details correctly on a failure with a .not: toFailTestAndMatchSnapshot 1`] = `
"expect(received).not.toFailWith(expected)
Expected:
  Not failure with \\"oops\\"
Received:
  Failure with \\"oops\\""
`;
